src/daemon/internal_proto.rs

```diff
@@ -0,0 +1,433 @@
+                        command: cmd.clone(),
+                        status: status.map(|i| i as i64),
+                        logs: logs.iter().map(|l| l.to_string()).collect(),
+                    }),
+                    output: None,
+                },
+                BuildError::Output { msg } => proto::Failure {
+                    kind: output,
+                    nix_file: nix_file.to_string(),
+                    io: None,
+                    spawn: None,
+                    exit: None,
+                    output: Some(proto::OutputFail {
+                        message: msg.clone(),
+                    }),
+                },
+            }),
+            _ => Err("expecting build_loop::Event::Failure"),
+        }
+    }
+}
+
+impl TryFrom<proto::Failure> for build_loop::Event {
+    type Error = String;
+
+    fn try_from(f: proto::Failure) -> Result<Self, Self::Error> {
+        Ok(build_loop::Event::Failure {
+            nix_file: f.nix_file.clone().into(),
+            failure: f.try_into()?,
+        })
+    }
+}
+
+impl TryFrom<proto::Failure> for error::BuildError {
+    type Error = &'static str;
+
+    fn try_from(pf: proto::Failure) -> Result<Self, Self::Error> {
+        use error::BuildError;
+        use proto::Failure_kind::*;
+
+        match pf {
+            proto::Failure {
+                kind: io,
+                io: Some(proto::IOFail { message }),
+                ..
+            } => Ok(BuildError::Io { msg: message }),
+            proto::Failure {
+                kind: spawn,
+                spawn: Some(proto::SpawnFail { command, message }),
+                ..
+            } => Ok(BuildError::Spawn {
+                cmd: command,
+                msg: message,
+            }),
+            proto::Failure {
+                kind: exit,
+                exit:
+                    Some(proto::ExitFail {
+                        command,
+                        status,
+                        logs,
+                    }),
+                ..
+            } => Ok(BuildError::Exit {
+                cmd: command,
+                status: status.map(|i| i as i32),
+                logs: logs.iter().map(|l| l.clone().into()).collect(),
+            }),
+            proto::Failure {
+                kind: output,
+                output: Some(proto::OutputFail { message }),
+                ..
+            } => Ok(BuildError::Output { msg: message }),
+            _ => Err("unexpected form of proto::Failure"),
+        }
+    }
+}
+
+impl TryFrom<&NixFile> for internal_proto::ShellNix {
+    type Error = &'static str;
+
+    fn try_from(nix_file: &NixFile) -> Result<Self, Self::Error> {
+        match nix_file {
+            NixFile::Shell(shell) => match shell.as_os_str().to_str() {
+                Some(s) => Ok(internal_proto::ShellNix {
+                    path: s.to_string(),
+                }),
+                None => Err("nix file path is not UTF-8 clean"),
```

Comment:
This looks like a bug to me, we shouldn’t require paths to be utf-8 clean, ever. Can varlink not transfer `[u8]`?

Reply:
"Can it not" is unclear. 

Lorri uses a `PathBuf`, but varlink needs to ship a `String` (because JSON). What I found to do that was `shell.as_os_str().to_str()` which returns `None` if there's unclean bytes. In proper operation, that `None` branch should never get exercised, I think? When would `OsStr` not be representable as `String`?